pdf-converter/client.py

```diff
@@ -0,0 +1,427 @@
+        check_range(untrusted_h, MAX_IMG_HEIGHT)
+    except ValueError as e:
+        logging.error(f"invalid image measurements received {e}")
+        raise DimensionError from e
+    else:
+        width = untrusted_w
+        height = untrusted_h
+
+    size = width * height * 3
+
+    return Dimensions(width=width, height=height, size=size, depth=DEPTH)
+
+
+###############################
+#         PDF-related
+###############################
+
+
+async def send_pdf(loop, proc, path):
+    try:
+        filesize = (await loop.run_in_executor(None, path.stat)).st_size
+        await send(proc, filesize)
+
+        data = await loop.run_in_executor(None, path.read_bytes)
+        await send(proc, data)
+    except BrokenPipeError:
+        raise
+
+
+async def recv_pagenums(loop, proc):
+    try:
+        untrusted_pagenums = int(await recvline(proc))
+    except (AttributeError, EOFError, UnicodeError, ValueError) as e:
+        raise ReceiveError from e
+
+    try:
+        check_range(untrusted_pagenums, MAX_PAGES)
+    except ValueError as e:
+        logging.error("invalid number of pages received")
+        raise PageError from e
+    else:
+        pagenums = untrusted_pagenums
+
+    return pagenums
+
+
+def archive(path):
+    archive_path = Path(ARCHIVE_PATH, path.name)
+    path.rename(archive_path)
+    print(f"Original PDF saved as: {archive_path}")
+
+
+###############################
+#    Representation-related
+###############################
+
+
+def get_rep(tmpdir, page, initial, final):
+    name = Path(tmpdir, str(page))
+    return Representation(initial=name.with_suffix(f".{initial}"),
+                          final=name.with_suffix(f".{final}"))
+
+
+def save_rep(rep, data, size):
+    if rep.initial.write_bytes(data) != size:
+        logging.error("inconsistent initial representation file size")
+        raise RepresentationError
+
+
+async def recv_rep(loop, proc, tmpdir, page):
+    """Receive initial representation from the server
+
+    :param proc: Qrexec process to read STDIN from
+    :param path: File path which will store the initial representation
+    """
+    rep = get_rep(tmpdir, page, "rgb", "png")
+
+    try:
+        dim = await get_img_dim(proc)
+        data = await recv_b(proc, dim.size)
+        await loop.run_in_executor(None, save_rep, rep, data, dim.size)
+    except EOFError as e:
+        raise ReceiveError from e
+    except (DimensionError, ReceiveError, RepresentationError):
+        raise
+
+    return rep, dim
+
+
+async def start_convert(rep, dim):
+    cmd = ["convert", "-size", f"{dim.width}x{dim.height}", "-depth",
+            f"{dim.depth}", f"rgb:{rep.initial}", f"png:{rep.final}"]
+
+    try:
+        proc = await asyncio.create_subprocess_exec(*cmd)
+        await wait_proc(proc)
+    except asyncio.CancelledError:
+        terminate_proc(proc)
+
+    return proc
+
+
+# TODO (?): Save->delete PNGs in loop to avoid storing all PNGs in memory.
```

Comment:
What do you think about doing our processing in batches? We can get 20 or so RGB files at a time, `await` their conversions, combine the PNGs into an intermediate PDF (or a prior one if we're not on the first batch), and repeat until we finish. We only get 20 or so temp files at a time that way.

The only problem is I feel bad for the event loop when we're waiting on the conversions and not really doing anything else... I want to get the next batch of RGB files. Oh, maybe callbacks? Something like whenever a conversion finishes, get the next RGB so we're always doing something.

Plus it's configurable

Reply:
You can use [queue](https://docs.python.org/3/library/asyncio-queue.html#queue) with limited size for this. Then, one function would receive rgb files and put a path into a queue (when it will reach queue limit, this will block until space is made), and another would get rgb files from the queue and convert to pdf. You'd need to start those two in parallel.

This makes the code slightly more complex to read, but should behave nicely and also you get trivial parallel processing limit.

Reply:
Fixed in 5c268a2 and 36453ae.